tests/callbacks/test_generate.py

```diff
@@ -0,0 +1,120 @@
+# Copyright 2022 MosaicML Composer authors
+# SPDX-License-Identifier: Apache-2.0
+
+import math
+from typing import Dict, Optional
+from unittest.mock import Mock
+
+import pytest
+import torch
+from packaging import version
+
+from composer.callbacks import Generate
+from composer.trainer import Trainer
+from composer.utils import dist
+from tests.common.datasets import dummy_gpt_lm_dataloader
+from tests.common.markers import device, world_size
+from tests.common.models import configure_tiny_gpt2_hf_model
+
+
+@device('cpu', 'gpu')
+@pytest.mark.parametrize('use_fsdp', [True, False])
+@world_size(1, 2)
+class TestGenerate():
+
+    def _check_test_params(self, device, world_size, use_fsdp) -> None:
+        if use_fsdp and version.parse(torch.__version__) < version.parse('1.13.0'):
+            pytest.skip('FSDP requires torch >= 1.13.0')
+        if device == 'cpu' and use_fsdp:
+            pytest.skip('FSDP is not supported on CPU.')
+        if world_size == 1 and use_fsdp:
+            pytest.xfail((
+                'Generation with world size 1 and FSDP fails with '
+                '`RuntimeError: The tensor has a non-zero number of elements, '
+                'but its data is not allocated yet. Caffe2 uses a lazy allocation, '
+                'so you will need to call mutable_data() or raw_mutable_data() to actually allocate memory.` '
+                'This issue is resolved with world size > 1 by a dummy call to forward (see HuggingFaceModel.dummy_forward_called), '
+                'but for some reason fails with world size 1.'))
+        if device == 'cpu' and world_size > 1:
+            pytest.xfail(
+                'GPT2 is not currently supported with DDP. See https://github.com/huggingface/transformers/issues/22482 for more details.'
+            )
+
+    def _fsdp_config(self, use_fsdp: bool) -> Optional[Dict]:
+        if use_fsdp:
+            return {
+                'sharding_strategy': 'FULL_SHARD',
+            }
+        return None
+
+    def _create_trainer(self, device, max_duration, use_fsdp, generate_cb: Optional[Generate] = None) -> Trainer:
+        return Trainer(model=configure_tiny_gpt2_hf_model(),
+                       train_dataloader=dummy_gpt_lm_dataloader(),
+                       device=device,
+                       max_duration=max_duration,
+                       callbacks=generate_cb,
+                       fsdp_config={'sharding_strategy': 'FULL_SHARD'} if use_fsdp else None)
```

Comment:
Guessing you meant to use the fsdp config helper function here?

Reply:
ahh good catch, actually was going to just delete the helper! originally made the helper bc I was calling it multiple times  across tests before the _create_trainer helper